keyboards/whitefox/whitefox.c

```diff
@@ -16,3 +16,79 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+};
+
+led_config_t g_led_config = {
+    {
+      // Key Matrix to LED Index
+      {      0,      1,      2,      3,      4,      5,      6,      7,      8 },
+      {      9,     10,     11,     12,     13,     14,     15,     16,     17 },
+      {     18,     19,     20,     21,     22,     23,     24,     25,     26 },
+      {     27,     28,     29,     30,     31,     32,     33,     34,     35 },
+      {     36,     37,     38,     39,     40,     41,     42,     43,     44 },
+      {     45,     46,     47,     48,     49,     50,     51,     52,     53 },
+      {     54,     55,     56,     57,     58,     59,     60,     61,     62 },
+      {     63,     64,     65,     66,     67,     68,     69,     70, NO_LED },
+    }, {
+      // LED Index to Physical Position
+      {   0,  0 }, {  15,  0 }, {  30,  0 }, {  45,  0 }, {  60,  0 }, {  75,  0 }, {  90,  0 }, { 105,  0 }, { 119,  0 }, { 134,  0 }, { 149,  0 }, { 164,  0 }, { 179,  0 }, { 194,  0 }, { 209,  0 }, { 224,  0 },
+      {   4, 16 }, {  22, 16 }, {  37, 16 }, {  52, 16 }, {  67, 16 }, {  82, 16 }, {  97, 16 }, { 112, 16 }, { 127, 16 }, { 142, 16 }, { 157, 16 }, { 172, 16 }, { 187, 16 }, { 205, 16 }, { 224, 16 },
+      {   6, 32 }, {  26, 32 }, {  41, 32 }, {  56, 32 }, {  71, 32 }, {  86, 32 }, { 101, 32 }, { 116, 32 }, { 131, 32 }, { 146, 32 }, { 161, 32 }, { 175, 32 }, { 190, 32 }, { 207, 32 }, { 224, 32 },
+      {   2, 48 }, {  19, 48 }, {  34, 48 }, {  49, 48 }, {  63, 48 }, {  78, 48 }, {  93, 48 }, { 108, 48 }, { 123, 48 }, { 138, 48 }, { 153, 48 }, { 168, 48 }, { 189, 48 }, { 209, 48 }, { 224, 48 },
+      {   2, 64 }, {  21, 64 }, {  39, 64 },                                        {  95, 64 },                           { 149, 64 }, { 164, 64 }, { 179, 64 }, { 194, 64 }, { 209, 64 }, { 224, 64 }
+    }, {
+      // LED Index to Flag
+      1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,
+      1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1,
+      1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1,
+      1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1, 1, 1,
+      1, 1, 1,          1,       1, 1, 1, 1, 1, 1,
+    }
+};
+#endif
+
+void keyboard_pre_init_kb(void) {
+#ifdef LED_MATRIX_ENABLE
+    // Turn on LED controller
+    setPinOutput(B16);
+    writePinHigh(B16);
+#endif
+    keyboard_pre_init_user();
+}
+
+void matrix_init_kb(void) {
+    // put your keyboard start-up code here
+    // runs once when the firmware starts up
+
+#ifdef LED_MATRIX_ENABLE
+    /*
+     * Since K20x is stuck with a 32 byte EEPROM (see tmk_core/common/chibios/eeprom_teensy.c),
+     * and neither led_matrix_eeconfig.speed or .flags fit in this boundary, just force their values to default on boot.
+     */
+#    if !defined(LED_MATRIX_STARTUP_SPD)
+#        define LED_MATRIX_STARTUP_SPD UINT8_MAX / 2
+#    endif
+    led_matrix_set_speed(LED_MATRIX_STARTUP_SPD),
+    led_matrix_set_flags(LED_FLAG_ALL);
+#endif
```

Comment:
Ideally, I would like to see if something like #12947 for the K20x fixes this, before merging.

Reply:
As far as my research has lead me to understand, you basically need to wipe the entire MCU and reflash the bootloader to change the "EEPROM" size. 🙁

Reply:
Then I'm not sure how else this can be solved if we want VIA support (and room for future core settings)... perhaps we should just fall back to the transient EEPROM implementation, that will eliminate the need for this code at least.

Reply:
I've done some more investigation and realized that the partitioning code in eeprom_teensy.c isn't actually run on an already partitioned MCU. When I tried increasing `EEPROM_SIZE` locally, I guess my board therefore just crashed due to the firmware configuration not matching the hardware configuration.

The documentation is a bit fuzzy on the subject, but NXP Support claimed in their forum that the relevant flash and registers need to be in "an erased state" before running the PGMPART command. I'm a bit hesitant to try since I don't want to risk bricking my main keyboard, but it might be possible to try skipping the [`FTFL->FCNFG & FTFL_FCNFG_RAMRDY` if statement](https://github.com/qmk/qmk_firmware/blob/master/tmk_core/common/chibios/eeprom_teensy.c#L99) and that way force the PGMPART command to be run. Due to what NXP Support said in the forums, my guess is that it won't work.

Reply:
> I've done some more investigation and realized that the partitioning code in eeprom_teensy.c isn't actually run on an already partitioned MCU.

As far as I can tell, the original Kiibohd firmware doesn't do any partitioning (so there is no persistent storage, apparently). This code seems to be mainly intended for actual Teensys, which have an external chip doing the bootloading, and according to the PJRC forum thread mentioned in the other PR, it sends the `SETRAM` command to turn the EEPROM functionality off. It also likely does a mass erase first.

And yeah, the manual says:
> After clearing CCIF to launch the Program Partition command, the flash memory module first verifies that the EEPROM Data Size Code and FlexNVM Partition Code in the data flash IFR are erased.

Reply:
> As far as I can tell, the original Kiibohd firmware doesn't do any partitioning (so there is no persistent storage, apparently).

Hmm, so it might have been QMK that set my keyboard to 32 bytes of EEPROM, then?

Reply:
Sooo, I finally mustered up the courage to try and force the partitioning command to run, and, as expected, it had no effect whatsoever (I did make sure the code actually ran, since `eeprom_initialize` isn't even called if the EEPROM is ready before the first read or write). Reading the manual, it seems like the only way to clear the data flash IFR (after which partitioning should be allowed again) is the "clear all blocks" command, which undoubtedly would remove the bootloader. Not really a safe move...

The transient EEPROM driver might be the way to go for non-Teensy K20x boards... 😢 

I guess crazy users (like me 😛) who _really_ want to use EEPROM could theoretically implement a custom (hackish) keymap level EEPROM driver that reshuffles the eeconfig addresses to more effectively use the 32 available bytes. \*evil laugh*

Reply:
At least the bootloaders for all Input:Club boards can be found here: https://github.com/kiibohd/controller/releases
`ergodox.v0.5.7.bootloader.bin`

All you need is a J-Link, I suppose.

Reply:
How should we go ahead with this? I feel like changing the boards to the transient EEPROM driver might be a bit out of scope for this PR, but without either that or these lines of code, LED Matrix is effectively dead after the second boot (due to `flags` being 0)...

Reply:
I'm fine to merge this as is, now that I know a bit more about the problem.

Since this is technically a bug in QMK, it should get a proper fix. I would kinda like to fix the code in a separate, breaking changes PR - it would mean users will need to wipe their boards, of course. But, it should only need to be done once.
Failing that, perhaps we can adapt the STM32 emulated eeprom?

Reply:
> Since this is technically a bug in QMK, it should get a proper fix. I would kinda like to fix the code in a separate, breaking changes PR - it would mean users will need to wipe their boards, of course. But, it should only need to be done once.

I have some doubts about how many users would be willing (and able) to wipe their boards like that...

If you just increase the default `EEPROM_SIZE` without changing anything else, boards initialized with another value will fail to boot. At least that's how mine behave.

Reply:
Hmm. It might even be possible to switch the FlexRAM function over from EEPROM to RAM when an incompatible EESIZE is detected in `SIM->FCFG1`, in effect making the EEPROM driver work using the same code, but behaving transient. I'm gonna do some experimenting. 👨‍🔬

Reply:
[Working prototype](https://github.com/firetech/qmk_firmware/commit/kinetis_flexram_fallback)! 😄

I also realized that the "crash" during boot after increasing `EEPROM_SIZE` was just an infinite loop in `flexram_wait()` due to the `EEERDY` flag never being set after an out-of-bounds write. Just making sure the FlexRAM isn't written to outside the partitioned EEPROM size is enough to make the board boot properly. In other words, the entire `if (eeprom_size < EEPROM_SIZE) { ... }` block _can_ be removed (but it might be desired).

If you feel that this would be a valid solution to the problem, I can make a PR of this after #12947 has been merged (would conflict heavily otherwise).

Reply:
An issue detailing all of these findings may be a good idea, rather than having it all in this tangentially-related PR discussion.

Reply:
Good point... I'll get one started.

Reply:
#13164